File: front/central/src/shared/utils/rate-pricing.ts
export interface PriceableRate {
    flete?: number;
    minimumInsurance?: number;
    extraInsurance?: number;
    cod?: boolean;
    codCarrierFee?: number;
    codProbabilityMargin?: number;
}

export interface RatePricingOptions {
    cod?: boolean;
    insured?: boolean;
}

const num = (value: unknown): number => {
    const n = Number(value ?? 0);
    return Number.isFinite(n) ? n : 0;
};

export const rateAppliesCOD = (rate: PriceableRate | null | undefined, opts: RatePricingOptions = {}): boolean =>
    Boolean(opts.cod) && rate?.cod === true;

export const rateGuideCost = (rate: PriceableRate | null | undefined, opts: RatePricingOptions = {}): number => {
    if (!rate) return 0;
    const insurance = num(rate.minimumInsurance) + (opts.insured ? num(rate.extraInsurance) : 0);
    const margin = rateAppliesCOD(rate, opts) ? num(rate.codProbabilityMargin) : 0;
    return num(rate.flete) + insurance + margin;
};

export const rateCarrierFee = (rate: PriceableRate | null | undefined, opts: RatePricingOptions = {}): number =>
    rateAppliesCOD(rate, opts) ? num(rate?.codCarrierFee) : 0;

export const rateTotalCost = (rate: PriceableRate | null | undefined, opts: RatePricingOptions = {}): number =>
    rateGuideCost(rate, opts) + rateCarrierFee(rate, opts);

export interface RateBreakdown {
    flete: number;
    minimumInsurance: number;
    extraInsurance: number;
    carrierFee: number;
    guideCost: number;
    total: number;
}

export const rateBreakdown = (rate: PriceableRate | null | undefined, opts: RatePricingOptions = {}): RateBreakdown => {
    const margin = rateAppliesCOD(rate, opts) ? num(rate?.codProbabilityMargin) : 0;
    const flete = num(rate?.flete) + margin;
    const minimumInsurance = num(rate?.minimumInsurance);
    const extraInsurance = opts.insured ? num(rate?.extraInsurance) : 0;
    const carrierFee = rateCarrierFee(rate, opts);
    const guideCost = flete + minimumInsurance + extraInsurance;
    return { flete, minimumInsurance, extraInsurance, carrierFee, guideCost, total: guideCost + carrierFee };
};

File: mobile/mobile_central/lib/shared/utils/rate_pricing.dart
import '../../services/modules/shipments/domain/entities.dart';

class RatePricingOptions {
  const RatePricingOptions({this.cod = false, this.insured = false});

  final bool cod;
  final bool insured;
}

class RateBreakdown {
  const RateBreakdown({
    required this.flete,
    required this.minimumInsurance,
    required this.extraInsurance,
    required this.carrierFee,
    required this.guideCost,
    required this.total,
  });

  final double flete;
  final double minimumInsurance;
  final double extraInsurance;
  final double carrierFee;
  final double guideCost;
  final double total;
}

class RatePricing {
  const RatePricing._();

  static double _num(double? value) {
    if (value == null || value.isNaN || value.isInfinite) return 0;
    return value;
  }

  static bool appliesCod(EnvioClickRate? rate, RatePricingOptions options) =>
      options.cod && rate?.cod == true;

  static double guideCost(EnvioClickRate? rate, RatePricingOptions options) {
    if (rate == null) return 0;
    final insurance = _num(rate.minimumInsurance) +
        (options.insured ? _num(rate.extraInsurance) : 0);
    final double margin =
        appliesCod(rate, options) ? _num(rate.codProbabilityMargin) : 0;
    return _num(rate.flete) + insurance + margin;
  }

  static double carrierFee(EnvioClickRate? rate, RatePricingOptions options) =>
      appliesCod(rate, options) ? _num(rate?.codCarrierFee) : 0;

  static double totalCost(EnvioClickRate? rate, RatePricingOptions options) =>
      guideCost(rate, options) + carrierFee(rate, options);

  static RateBreakdown breakdown(EnvioClickRate? rate, RatePricingOptions options) {
    final double margin =
        appliesCod(rate, options) ? _num(rate?.codProbabilityMargin) : 0;
    final flete = _num(rate?.flete) + margin;
    final minimumInsurance = _num(rate?.minimumInsurance);
    final double extraInsurance =
        options.insured ? _num(rate?.extraInsurance) : 0;
    final fee = carrierFee(rate, options);
    final cost = flete + minimumInsurance + extraInsurance;
    return RateBreakdown(
      flete: flete,
      minimumInsurance: minimumInsurance,
      extraInsurance: extraInsurance,
      carrierFee: fee,
      guideCost: cost,
      total: cost + fee,
    );
  }
}
